File: MultiBoxClientJavaCore/src/sk/hackcraft/multibox/net/transformers/GetLibraryItemDecoder.java
package sk.hackcraft.multibox.net.transformers;

import sk.hackcraft.multibox.model.GenericLibraryItem;
import sk.hackcraft.multibox.model.LibraryItem;
import sk.hackcraft.multibox.model.LibraryItemType;
import sk.hackcraft.multibox.model.libraryitems.DirectoryItem;
import sk.hackcraft.multibox.model.libraryitems.MultimediaItem;
import sk.hackcraft.multibox.net.data.GetLibraryItemData;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public class GetLibraryItemDecoder extends JacksonMessageDecoder<GetLibraryItemData>
{
	@Override
	public GetLibraryItemData decodeJson(ObjectMapper objectMapper, String jsonString) throws Exception
	{
		JsonNode rootNode = objectMapper.readTree(jsonString);

		try
		{
			long id = rootNode.path("id").asLong();
			LibraryItemType type = LibraryItemType.valueOf(rootNode.path("type").asText());
			String name = rootNode.path("name").asText();
		
			LibraryItem libraryItem;
			switch (type)
			{
				case DIRECTORY:
					libraryItem = createDirectory(id, name, rootNode);
					break;
				case MULTIMEDIA:
					libraryItem = createMultimedia(id, name, rootNode);
					break;
				default:
					throw new IllegalArgumentException("Invalid library item type received from server.");
			}
			
			return new GetLibraryItemData(libraryItem);
		}
		catch (IllegalArgumentException e)
		{
			throw new RuntimeException(e);
		}
	}

	private LibraryItem createMultimedia(long id, String name, JsonNode rootNode)
	{
		int length = rootNode.path("length").asInt();
		
		return new MultimediaItem(id, name, length);
	}

	private LibraryItem createDirectory(long id, String name, JsonNode rootNode)
	{
		DirectoryItem directory = new DirectoryItem(id, name);
		
		for (JsonNode node : rootNode.path("items"))
		{
			long itemId = node.path("id").asLong();
			LibraryItemType itemType = LibraryItemType.valueOf(node.path("type").asText());
			String itemName = rootNode.path("name").asText();
			
			GenericLibraryItem libraryItem = new GenericLibraryItem(itemId, itemType, itemName);
			directory.addItem(libraryItem);
		}
		
		return directory;
	}
}

File: MultiboxClientCSharpCore/Net/Transformer/GetLibraryItemDecoder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using CSharpNetInterface.Message;
using MultiboxClientCSharpCore.Model;
using MultiboxClientCSharpCore.Model.LibraryItems;
using MultiboxClientCSharpCore.Net.Data;
using MultiboxClientCSharpCore.Net.Transformer.Json;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace MultiboxClientCSharpCore.Net.Transformer
{
	public class GetLibraryItemDecoder : IDataTransformer<string, GetLibraryItemResultData>
	{
		public GetLibraryItemResultData Transform(string input)
		{
			JObject rootNode = JObject.Parse(input);

			JObject libraryItemNode = rootNode["libraryItem"] as JObject;

			long id = libraryItemNode["id"].ToObject<long>();
			LibraryItemType type = LibraryItemTypeParser.Parse(libraryItemNode["type"].ToObject<string>());
			string name = libraryItemNode["name"].ToObject<string>();

			ILibraryItem libraryItem;
			switch (type)
			{
				case LibraryItemType.Directory:
					libraryItem = CreateDirectory(id, name, libraryItemNode);
					break;
				case LibraryItemType.Multimedia:
					libraryItem = CreateMultimedia(id, name, libraryItemNode);
					break;
				default:
					throw new ArgumentException("Invalid library item type received from server.");
			}

			return new GetLibraryItemResultData(libraryItem);
		}

		private ILibraryItem CreateMultimedia(long id, String name, JObject rootNode)
		{
			int length = rootNode["length"].ToObject<int>();

			return new MultimediaItem(id, name, length);
		}

		private ILibraryItem CreateDirectory(long id, String name, JObject rootNode)
		{
			DirectoryItem directory = new DirectoryItem(id, name);
		
			foreach (JObject node in (rootNode["items"] as JArray))
			{
				long itemId = node["id"].ToObject<long>();
				LibraryItemType itemType = LibraryItemTypeParser.Parse(node["type"].ToObject<string>());
				string itemName = node["name"].ToObject<string>();
			
				GenericLibraryItem libraryItem = new GenericLibraryItem(itemId, itemType, itemName);
				directory.AddItem(libraryItem);
			}
		
			return directory;
		}
	}
}
